### PR TITLE
CheckRightsService_RemoveRightsFromUser

### PR DESCRIPTION
**Pull request by KostinP**
I started from https://github.com/LT-Students/DigitalOffice/pull/49 pull request so that there were fewer conflicts. Didn't make a repository, it doesn't work
(You also need to add controller tests)

### CODE REVIEW

File: DigitalOffice/Services/CheckRightsService/Validator/RemoveRightsFromUserValidator.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        public RemoveRightsFromUserValidator()
+        {
+            RuleFor(rights => rights.UserId)
+                .NotEmpty()
+                .WithName("User Id");
+
+            RuleFor(rights => rights.RightsId)
+                .NotEmpty()
+                .WithName("Right Id");
+        }
```
Comment (denis-kriv) [inline DigitalOffice/Services/CheckRightsService/Validator/RemoveRightsFromUserValidator.cs]: I am not sure, but maybe we have more rules.

File: DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs [hunk 2]
```diff
@@ -14,5 +14,13 @@ public List<Right> GetRightsList([FromServices] IGetRightsListCommand command)
         {
             return command.Execute();
         }
+
+        [HttpPost("removeRightsFromUser")]
```
Comment (denis-kriv) [inline DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs]: Why not to use `HttpDelete`

File: DigitalOffice/Tests/CheckRightsServiceTests/Validators/RemoveRightsFromUserValidatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using System.Collections.Generic;
+
+namespace LT.DigitalOffice.CheckRightsServiceUnitTests.Validators
+{
+    public class RemoveRightsFromUserValidatorTests
+    {
+        private IValidator<RemoveRightsFromUserRequest> validator;
+
+        [SetUp]
+        public void Initialize()
```
Comment (denis-kriv) [inline DigitalOffice/Tests/CheckRightsServiceTests/Validators/RemoveRightsFromUserValidatorTests.cs]: `SetUp`

Comment (KostinP) [reply DigitalOffice/Tests/CheckRightsServiceTests/Validators/RemoveRightsFromUserValidatorTests.cs]: Fixed.

File: DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs (hunk 2 above)
Comment (KostinP) [reply DigitalOffice/Services/CheckRightsService/Controllers/CheckRightsController.cs]: Fixed.

File: DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs [hunk 4]
```diff
@@ -23,5 +25,37 @@ public List<Right> GetRightsList()
...
+        public bool RemoveRightsFromUser(RemoveRightsFromUserRequest request)
+        {
+            foreach (var rightId in request.RightsId)
+            {
+                var dbRight = dbContext.Rights.Include(r => r.UserIds).FirstOrDefault(right => right.Id == rightId);
+                if (dbRight == null)
+                {
+                    // TODO: add custom exception
+                    throw new Exception("Right doesn't exist.");
+                }
```
Comment (DarkFelas) [inline DigitalOffice/Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: remove

**Comment by KostinP** [comment]
Problem: when creating a new migration, it still tries to delete data.
@mekk1t @DarkFelas 

Perhaps Nikita did not have this, because he did not touch the file CheckRightsServiceDbContextModelSnapshot.cs that is responsible for the current state of the model stored. It enables the migrations framework to calculate the changes required to bring the database up to date with the model.

What to do?

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IRemoveRightsFromUserCommand.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    /// Provides method for removing rights from user.
+    /// </summary>
+    public interface IRemoveRightsFromUserCommand
+    {
+        /// <summary>
+        /// Remove rights from user.
+        /// </summary>
+        /// <param name="request">Request with rights and user id.</param>
+        /// <returns>Return true if successfully else return false.</returns>
+        bool Execute(RemoveRightsFromUserRequest request);
```
Comment (DarkFelas) [inline DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IRemoveRightsFromUserCommand.cs]: no need for bool. just **void**

**Review by DarkFelas — APPROVED** [review]
Does not saw any problem with migration, all works well.

File: DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IRemoveRightsFromUserCommand.cs (hunk 5 above)
Comment (KostinP) [reply DigitalOffice/Services/CheckRightsService/Commands/Interfaces/IRemoveRightsFromUserCommand.cs]: Fixed.

**Review by lalagvana — APPROVED** [review]

File: Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs [hunk 6]
```diff
@@ -33,22 +75,96 @@ public void Clear()
...
+            Assert.AreEqual(rightsBeforeRequest, rightsAfterRequest);
+            // Removed required rights.
+            userRightsBeforeRequest.RemoveAll(ru =>
+                ru.UserId == request.UserId && request.RightIds.Contains(ru.RightId));
+            Assert.AreEqual(userRightsBeforeRequest, userRightsAfterRequest);
+
+        }
+
+        [Test]
+        public void ShouldNotDeleteAnythingWhenRightIdIsNoFound()
```
Comment (lalagvana) [inline Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs]: IsNotFound

File: DigitalOffice/Services/CheckRightsService/Validator/RemoveRightsFromUserValidator.cs (hunk 1 above)
Comment (denis-kriv) [reply DigitalOffice/Services/CheckRightsService/Validator/RemoveRightsFromUserValidator.cs]: have to add*

File: Services/CheckRightsService/Commands/RemoveRightsFromUserCommand.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            this.validator = validator;
+        }
+
+        public void Execute(RemoveRightsFromUserRequest request)
+        {
+            validator.ValidateAndThrow(request);
+            repository.RemoveRightsFromUser(request);
+
+            // TODO: Check rights of user (sender) with Kernel_AccessValidator
+        }
```
Comment (djino97) [inline Services/CheckRightsService/Commands/RemoveRightsFromUserCommand.cs]: Will you add check rights in the this pr?

**Review by mekk1t — APPROVED** [review]
Add AccessValidator

**Review by denis-kriv — APPROVED** [review]
Add some rules for validator please

**Review by DarkFelas — APPROVED** [review]

File: Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs (hunk 6 above)
Comment (KostinP) [reply Tests/CheckRightsServiceTests/Repositories/CheckRightsRepositoryTests.cs]: Fixed.

File: Services/CheckRightsService/Commands/RemoveRightsFromUserCommand.cs (hunk 7 above)
Comment (KostinP) [reply Services/CheckRightsService/Commands/RemoveRightsFromUserCommand.cs]: Fixed.

**Pull request merged**